### DISCUSSION TITLE
如何修改 `StateParam` 以使其支持自定义类型的标注？

### DISCUSSION THREAD

**Discussion by A-kirami** [nonebot/discussions#297] [问答] [answered]
```py
@matcher.handle()
async def test(state: State) -> None:
    print(state)
    print(type(state))
```

`State` 是 `T_State` 的子类，但在测试中，`state` 类型却是 `dict`，应该如何修改 `StateParam` 才能使其类型正确？
还是说应该修改其他部分？

**Comment by yanyongyu** [comment]
state就是nb提供的state，不会改变类型，如果你需要一个自定义的类型，可以通过依赖注入来返回

```python
from nonebot.param import Depends
from pydantic import BaseModel, Extra

class Data(BaseModel, extra=Extra.ignore):
	xxx: int = 1

def dict_to_model(state: T_State) -> Data:
	return Data.parse_obj(state)

@matcher.handle()
async def test(state: Data = Depends(dict_to_model)):
	assert type(state) is Data
```

**Reply by A-kirami** [reply]
> state就是nb提供的state，不会改变类型，如果你需要一个自定义的类型，可以通过依赖注入来返回
> 
> ```python
> from nonebot.param import Depends
> from pydantic import BaseModel, Extra
> 
> class Data(BaseModel, extra=Extra.ignore):
> 	xxx: int = 1
> 
> def dict_to_model(state: T_State) -> Data:
> 	return Data.parse_obj(state)
> 
> @matcher.handle()
> async def test(state: Data = Depends(dict_to_model)):
> 	assert type(state) is Data
> ```

并非是直接修改 nb，而是想在基于 nb 开发的 bot 通过🐒补丁实现这一特性。
使用依赖注入有些繁琐，更希望直接在 bot 框架中提供支持。

**Comment by A-kirami** [comment, marked as answer]
已解决，除了 `StateParam`，还需要修改 `Matcher.new()` 方法。
